File: problems/0301-0400/0363_capped-submatrix-sum/solution_sorted_prefix_search.go
import "sort"

func cappedSubmatrixSum(matrix [][]int, k int) int {
	m := len(matrix)
	n := len(matrix[0])
	best := 0
	found := false
	for top := 0; top < m; top++ {
		// colSum[c] = sum of column c between rows top..bottom, so
		// extending the bottom row is one O(n) update; any rectangle
		// in this row pair is a contiguous subarray of colSum.
		colSum := make([]int, n)
		for bottom := top; bottom < m; bottom++ {
			for c := 0; c < n; c++ {
				colSum[c] += matrix[bottom][c]
			}
			prefix := 0
			// 0 seeded so a subarray starting at the first column counts.
			prefixes := []int{0}
			for c := 0; c < n; c++ {
				prefix += colSum[c]
				// Subarray sum = prefix - earlier prefix; the smallest
				// earlier >= prefix-k maximizes it while staying <= k.
				// bisect_left(prefixes, prefix-k)
				lo, hi := 0, len(prefixes)
				for lo < hi {
					mid := (lo + hi) / 2
					if prefixes[mid] < prefix-k {
						lo = mid + 1
					} else {
						hi = mid
					}
				}
				if lo < len(prefixes) {
					candidate := prefix - prefixes[lo]
					if !found || candidate > best {
						best = candidate
						found = true
					}
				}
				// insort keeps the list sorted for the next query.
				pos := sort.SearchInts(prefixes, prefix)
				prefixes = append(prefixes, 0)
				copy(prefixes[pos+1:], prefixes[pos:])
				prefixes[pos] = prefix
			}
		}
	}
	return best
}

File: problems-originals/0301-0400/0363_max-sum-of-rectangle-no-larger-than-k/solution_sorted_prefix_search.rs
impl Solution {
    pub fn max_sum_submatrix(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
        let m = matrix.len();
        let n = matrix[0].len();
        let k = k as i64;
        let mut best: Option<i64> = None;
        for top in 0..m {
            // col_sum[c] = sum of column c between rows top..bottom, so
            // extending the bottom row is one O(n) update; any rectangle
            // in this row pair is a contiguous subarray of col_sum.
            let mut col_sum = vec![0i64; n];
            for bottom in top..m {
                for c in 0..n {
                    col_sum[c] += matrix[bottom][c] as i64;
                }
                let mut prefix: i64 = 0;
                // 0 seeded so a subarray starting at the first column counts.
                let mut prefixes: Vec<i64> = vec![0];
                for c in 0..n {
                    prefix += col_sum[c];
                    // Subarray sum = prefix - earlier prefix; the smallest
                    // earlier >= prefix - k maximizes it while staying <= k.
                    let target = prefix - k;
                    let position = prefixes.partition_point(|&x| x < target);
                    if position < prefixes.len() {
                        let candidate = prefix - prefixes[position];
                        if best.is_none() || candidate > best.unwrap() {
                            best = Some(candidate);
                        }
                    }
                    // Keep the list sorted for the next query.
                    let at = prefixes.partition_point(|&x| x < prefix);
                    prefixes.insert(at, prefix);
                }
            }
        }
        best.unwrap() as i32
    }
}
